shell/platform/embedder/embedder.cc

```diff
@@ -624,6 +727,59 @@ static sk_sp<SkSurface> MakeSkSurfaceFromBackingStore(
 #endif
 }
 
+static sk_sp<SkSurface> MakeSkSurfaceFromBackingStore(
+    GrDirectContext* context,
+    const FlutterBackingStoreConfig& config,
+    const FlutterVulkanBackingStore* vulkan) {
+#ifdef SHELL_ENABLE_VULKAN
+  if (!vulkan->image.image) {
+    FML_LOG(ERROR) << "Embedder supplied null Vulkan image.";
+    return nullptr;
+  }
+
+  GrVkImageInfo image_info = {
```

Comment:
A lot of the flags here are completely different from what Fuchsia is using, I had to hack this section up a lot to get it working....  See https://github.com/flutter/engine/blob/main/shell/platform/fuchsia/flutter/vulkan_surface.cc#L364

In particular we make use of the following properties:
+VkMemory handle to the memory which backs the VkImage
+Allocation size of the VkMemory via VkImageMemoryRequirements
+Tiling, InitialLayout, Format, MipLevels, and Usage from the VkImageCreateInfo

It looks like the Tiling, InitialLayout, Format, and Usage happen to match between Fuchsia code and this....but I'm not sure if we should rely on that in general; if the Embedder calls VkCreateImage with different flags it will probably fail horribly.

I had to modify the FlutterVulkanBackingStore to contain these fields to get it working for Fuchsia

Reply:
Ah, I definitely agree the tiling/layout/format/mipmap config/usage flags should be provided through the FlutterVulkanBackingStore (and perhaps a validation should be added to enforce that "VK_IMAGE_USAGE_COLOR_ATTACHMENT_BIT" is set), but I don't think embedders should need to supply VkMemory handles over the API barrier -- the embedder should be responsible for binding the VkImage to memory before providing it to the API via the callback.

In your hacked up version, could you try creating the `GrVkImageInfo` here without providing any memory/alloc info and passing it to `SkSurface::MakeFromBackendTexture`? If all the other config is correct, it should still work without memory info as it does in the test compositor.

Reply:
On second thought, this configuration is mostly the concern of internal image view and sampler creation. I looked into each one in more detail. TL;DR: Having embedders supply the `Format` is probably necessary, but having embedders supply the `ImageLayout`/`MipLevels`/`Usage` isn't, although the required usage options should be explicitly documented.
* `Format`: The `Format` is certainly consequential for Skia to correctly attach the color component in the framebuffer for rendering, so having this passed in is probably a good idea. I'm gonna look into what our texture restrictions are to figure out what formats can be supported. My immediate guess is that we're internally limited to swizzle variations of RGBA8.
* `ImageLayout`: The first thing Flutter does with this embedder-provided image is attach it to a framebuffer and overwrite all its contents, so we shouldn't need to worry about doing a layout transition (presumably to then sample the previous contents of the image). And so it should be safe to set `ImageLayout` to `VK_IMAGE_LAYOUT_UNDEFINED` here.
* `MipLevels`: Unless Flutter submits commands to blit MIPs after rendering these layers, supplying a MIP count here would go unused. If Fuchsia has a use case for LoD sampling these layers in its compositor in the future, Fuchsia's embedder can still supply multi-MIP images here and Flutter will just render into MIP 0. Fuchsia's embedder could then just submit commands to blit the remaining MIPs (with full control over filtering) as the layers are handed back.
* `Usage`: There would be no purpose for receiving the `Usage` from the embedder since there is no circumstance where these usage flags need to exactly match the usage flags that were supplied when creating the image. Per the spec for e.g. VkImageViews:
  > image view creation inherits the implicit parameter usage specifying the allowed usages of the image view that, by default, takes the value of the corresponding usage parameter specified in VkImageCreateInfo at image creation time. The implicit usage can be overriden by adding a VkImageViewUsageCreateInfo structure to the pNext chain, but the view usage must be a subset of the image usage.
  
  So we should just set the usage flags to be the minimal set of things necessary here and document required flags in the API. For example, the embedder must _always_ create the image with `VK_IMAGE_USAGE_COLOR_ATTACHMENT_BIT`, but the rest of these flags are most likely not actually required here.

Reply:
Adding format (and color space as mentioned below) + documenting Usage SGTM.

I'll try that out in my hacky change; I don't anticipate any difficulties

Reply:
Alrighty, I added the format to FlutterVulkanImage.